src/Microsoft.FeatureManagement/Targeting/ContextualTargetingFilter.cs

```diff
@@ -0,0 +1,186 @@
+        /// </summary>
+        /// <param name="contextId">A context identifier that determines what the percentage is applicable for</param>
+        /// <param name="percentage">The total percentage of possible context identifiers that should be targeted</param>
+        /// <returns>A boolean representing if the context identifier should be targeted</returns>
+        private bool IsTargeted(string contextId, double percentage)
+        {
+            byte[] hash;
+
+            using (HashAlgorithm hashAlgorithm = SHA256.Create())
+            {
+                hash = hashAlgorithm.ComputeHash(Encoding.UTF8.GetBytes(contextId));
+            }
+
+            //
+            // Use first 4 bytes for percentage calculation
+            // Cryptographic hashing algorithms ensure adequate entropy across hash
+            uint contextMarker = BitConverter.ToUInt32(hash, 0);
+
+            double contextPercentage = (contextMarker / (double)uint.MaxValue) * 100;
+
+            return contextPercentage < percentage;
+        }
+
+        /// <summary>
+        /// Performs validation of targeting settings.
+        /// </summary>
+        /// <param name="settings">The settings to validate.</param>
+        /// <param name="paramName">The name of the invalid setting, if any.</param>
+        /// <param name="reason">The reason that the setting is invalid.</param>
+        /// <returns>True if the provided settings are valid. False if the provided settings are invalid.</returns>
+        private bool TryValidateSettings(TargetingFilterSettings settings, out string paramName, out string reason)
+        {
+            const string OutOfRange = "The value is out of the accepted range.";
+
+            const string RequiredParameter = "Value cannot be null.";
+
+            paramName = null;
+
+            reason = null;
+
+            if (settings.Audience == null)
+            {
+                paramName = nameof(settings.Audience);
+
+                reason = RequiredParameter;
+
+                return false;
+            }
+
+            if (settings.Audience.DefaultRolloutPercentage < 0 || settings.Audience.DefaultRolloutPercentage > 100)
+            {
+                paramName = $"{settings.Audience}.{settings.Audience.DefaultRolloutPercentage}";
+
+                reason = OutOfRange;
+
+                return false;
+            }
+
+            if (settings.Audience.Groups != null)
+            {
+                int index = 0;
+
+                foreach (GroupRollout groupRollout in settings.Audience.Groups)
+                {
+                    index++;
+
+                    if (groupRollout.RolloutPercentage < 0 || groupRollout.RolloutPercentage > 100)
+                    {
+                        //
+                        // Audience.Groups[1].RolloutPercentage
+                        paramName = $"{settings.Audience}.{settings.Audience.Groups}[{index}].{groupRollout.RolloutPercentage}";
```

Comment:
NIT: it may be easier for users debugging if we include `groupRollout.Name` in the message.

Reply:
@zhenlan It's a been more than I would expect in an argument exception. I'd be fine without it, but if you want it I'm not opposed. Let me know.

Reply:
It's okay.